### PR TITLE
Refactor `connection.rs` to make `fail_with` errors impossible

### LINKED ISSUE #1599: Panic: called fail_with on already-failed connection state, from msg=GetData
## Duplicate

This panic will be fixed by the fail_with refactor - so this ticket is a duplicate of #1610.

## Motivation

These panics are related to https://github.com/ZcashFoundation/zebra/issues/1510 ~which should have been fixed with~ which was partly fixed by https://github.com/ZcashFoundation/zebra/pull/1531.

Edit: we knew #1531 fixed some potential causes of #1510, but we weren't sure if we'd found them all

Issue report 1 is from commit 8a7b023.
Issue report 2 is from commit 1edb379e (#1620 + #1622).
The backtraces don't contain any useful information.

## Analysis 2

1. A peer calls `GetHeaders` then `GetData` (get blocks) on the same connection.
2. `GetHeaders` (or some intervening request) fails with "Connection reset by peer".
3. Zebra panics because processing the `GetData` request fails with a "Broken pipe" error.

When the `GetHeaders` request fails, Zebra should stop processing any further requests from that peer. Instead, it continues to try to process inbound messages from that peer.

## Analysis 1

Possibly resolved in #1600.

This panic can happen in two different ways:
1. The inbound service calls `drive_peer_request`, which doesn't check for a failed connection state before performing tasks that can potentially call `fail_with`.
2. The connection code calls async functions after checking for a failed connection state, so it's possible that the connection can fail between the check and a subsequent call to `fail_with`

So I suggest that we add the missing failed connection check, downgrade the panic to a warning, and add additional debugging information.

## Error 1
```
called fail_with on already-failed connection state
```

## Metadata 1
|key|value|
|--|--|
|**version**|1.0.0-alpha.0|
|**git commit**|8a7b023|
|**Zcash network**|Mainnet|
|**location**|/zebra/zebra-network/src/peer/connection.rs:515:13|

## SpanTrace 1

<details>

```
SpanTrace:
   0: zebra_network::peer::connection::msg_as_req
           with msg=getdata
             at zebra-network/src/peer/connection.rs:728
   1: zebra_network::peer::handshake::peer
           with addr=218.8.47.71:8233
             at zebra-network/src/peer/handshake.rs:198
   2: zebrad::application::
           with zebrad="8a7b023" net="Main"
             at zebrad/src/application.rs:254
```
</details>

## Error 2
```
calling fail_with on already-failed connection state: failed connections must stop processing pending requests and responses, then close the connection. state: Failed original error: SharedPeerError(Serialization(Io(Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }))) new error: SharedPeerError(Serialization(Io(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }))) client receiver: ClientRequestReceiver { inner: Receiver { inner: Some(BoundedInner { buffer: 0, state: 0, message_queue: Queue { head: 0x7fdcd4830060, tail: UnsafeCell }, parked_queue: Queue { head: 0x7fdcb41f21f0, tail: UnsafeCell }, num_senders: 2, recv_task: AtomicWaker }) } }
```

## Metadata 2
|key|value|
|--|--|
|**version**|1.0.0-alpha.1|
|**git commit**|1edb379e|
|**Zcash network**|Mainnet|
|**location**|/home/dev/zebra/zebra-network/src/peer/connection.rs:535:13|

## SpanTrace 2

<details>

```
SpanTrace:
   0: zebra_network::peer::connection::msg_as_req
           with msg=getdata
             at zebra-network/src/peer/connection.rs:762
   1: zebra_network::peer::handshake::peer
           with addr=163.172.222.30:8233
             at zebra-network/src/peer/handshake.rs:198
   2: zebrad::application::
           with zebrad="1edb379e" net="Main"
             at zebrad/src/application.rs:254
```
</details>

## Logs 2

<details>

```
Jan 30 07:09:34.078  INFO {zebrad="1edb379e" net="Main"}:sync:obtain_tips: zebra_state::service: pruned utxo requests old_len=4 new_len=0 prune_count=4 tip=Some((Height(1129116), block::Hash("0000000002121f23ad2a4fa88aa1c700e0d0535354d0e8d477f92e454d5fdf81")))
Jan 30 07:09:34.078  INFO {zebrad="1edb379e" net="Main"}:sync:obtain_tips:state: zebra_state::util: created block locator tip_height=Height(1129116) min_locator_height=1129017 locators=[Height(1129116), Height(1129115), Height(1129114), Height(1129112), Height(1129108), Height(1129100), Height(1129084), Height(1129052), Height(1129017)]
Jan 30 07:09:34.330  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: exhausted prospective tip set
Jan 30 07:09:34.330  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: waiting to restart sync timeout=61s
Jan 30 07:10:35.331  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: starting sync, obtaining new tips
Jan 30 07:10:35.332  INFO {zebrad="1edb379e" net="Main"}:sync:obtain_tips:state: zebra_state::util: created block locator tip_height=Height(1129116) min_locator_height=1129017 locators=[Height(1129116), Height(1129115), Height(1129114), Height(1129112), Height(1129108), Height(1129100), Height(1129084), Height(1129052), Height(1129017)]
Jan 30 07:10:41.332  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: exhausted prospective tip set                                                                                     
Jan 30 07:10:41.332  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: waiting to restart sync timeout=61s
Jan 30 07:11:42.333  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: starting sync, obtaining new tips
Jan 30 07:11:42.334  INFO {zebrad="1edb379e" net="Main"}:sync:obtain_tips:state: zebra_state::util: created block locator tip_height=Height(1129116) min_locator_height=1129017 locators=[Height(1129116), Height(1129115), Height(1129114), Height(1129112), Height(1129108), Height(1129100), Height(1129084), Height(1129052), Height(1129017)]
Jan 30 07:11:48.335  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: extending tips tips.len=1 in_flight=3 lookahead_limit=2000
Jan 30 07:11:54.337  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: exhausted prospective tip set                                                                                     
Jan 30 07:11:54.337  INFO {zebrad="1edb379e" net="Main"}:sync: zebrad::components::sync: waiting to restart sync timeout=61s
Jan 30 07:12:19.657  INFO {zebrad="1edb379e" net="Main"}:peer{addr=163.172.222.30:8233}:msg_as_req{msg=getheaders}: zebra_state::service: pruned utxo requests old_len=7 new_len=5 prune_count=2 tip=Some((Height(1129116), block::Hash("0000000002121f23ad2a4fa88aa1c700e0d0535354d0e8d477f92e454d5fdf81")))   
Jan 30 07:12:19.658  INFO {zebrad="1edb379e" net="Main"}:peer{addr=163.172.222.30:8233}:msg_as_req{msg=getheaders}:state: zebra_state::service: responding to peer GetBlocks or GetHeaders final_height=Height(1129116) response_len=20 chain_tip_height=Height(1129116) stop_height=None intersection_height=Some(Height(1129096))
The application panicked (crashed).
Message:  calling fail_with on already-failed connection state ...
```

</details>

### LINKED ISSUE #1610: Refactor the peer connection code to avoid failure handling bugs
## TODO

- [ ] work out why the refactor sends more errors to the syncer
- [ ] merge these changes
- [ ] follow-up by making this code a lot easier to maintain - split functions, map complex matches to custom enums

## Existing Work

PR https://github.com/ZcashFoundation/zebra/pull/1817

**Is your feature request related to a problem? Please describe.**

In #1531 and #1600, we fixed some failure handling bugs in the zebra-network peer connection code.

But the way the code is structured risks introducing more of these bugs in future.

We should refactor this code, to make these bugs impossible.

**Describe the solution you'd like**

* make fail_with take an owned Connection struct
* if other functions also need to take an owned struct, they should return ownership on success, and call fail_with on error
* split the failure cleanup code out of the task loop into its own separate function
* at the end of fail_with, call the cleanup code, and end the task
* abolish the Failed state, and delete or refactor all the code that uses it

**Describe alternatives you've considered**

We could do nothing, but that risks future bugs when developers add new failure modes, but don't actually test those failure modes.

We could create tests for every possible error path, but that's going to take ongoing effort.

**Additional context**

The panic bugs caused by this badly structured code are #1510 and #1599

### PR DESCRIPTION
**Pull request by yaahc**
## Motivation

Prior to this PR, we've been experiencing intermittent panics related to runtime invariants within `Connection` being invalidated. We've resolved most of these bugs but the ones related to `fail_with` have been stubborn enough that we decided it would be best to refactor the associated error handling logic to hopefully prevent such errors at compile time.

## Solution

This PR attempts to solve the `fail_with` panics by removing the API entirely, and replacing it with error propagation through return values. We refactored the primary event loop in `run` by extracting the body into a free function that _must_ produce a `Transition` struct describing what state transition the `Connection` should next take. This mechanism is now used instead of `fail_with` to propagate errors back to our callers and to cleanly tear down state when a `Connection` encounters an error. When we wish to close a connection we use `Transition::Close*` which cannot be used to construct a subsequent state instead of using `fail_with` to mutate some shared state which propagates back the error.

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Manual Integration Testing

## Review

@teor2345 

## Related Issues

Closes #1599 
Closes #1610

### CODE REVIEW

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs]: Copying this code feels like an anti-pattern - and it's a reversion of recent changes to this file.
Can we handle the oneshot request Nil responses in a different way?

For example, the peer routing for the Advertise* requests is handled by passing them to different functions:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/peer_set/set.rs#L469

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:755]: The goal of this removed match statement is to make sure that tx and Failed are used correctly.
Where are these checks moved to in this PR?

**Review by teor2345 — COMMENTED** [review]
This looks really good!

I just have a few questions about the oneshot sender checks and handling.

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [reply zebra-network/src/peer/connection.rs:755]: they shouldn't be necessary anymore because I factor'ed out the Option around the `tx` and documented the invariant that this function is only ever called from the `AwaitingRequest` state

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [reply zebra-network/src/peer/connection.rs]: making it so I didn't have to handle propagating the `tx` back out of these cases was pretty critical for removing the need to handle cases that shouldn't ever exist in practice. We can factor this out into a function but I don't see the point, it's one line of code, splitting it into a function will only add lines without reducing complexity.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:755]: I think I'm going to need a walkthrough and updated doc comments to understand this code.
(Feel free to defer the doc comments until the design has stabilised.)

To be clear, my goal is to make sure that:
1. every request sends a response on `tx`, or we have removed that invariant
2. requests that need responses actually send them, and requests that don't have responses don't send any

I'm not particularly concerned about code duplication.

I created a Request/Response design in #1723 that encodes request timing, response, and routing as types. I'm happy to leave that redesign until later, because it's not causing any bugs right now. But let me know if you'd like me to walk you through it - it might help explain why we treat some requests differently.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs]: It might be clearer to say `Close` rather than `Exit` here - typically, we close connections, and exit processes.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs]: Can we just drop all these requests, or do we need to return errors on them?

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs]: Can we move this common code into the transition function, so every message that doesn't expect a response gets the same treatment?

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:755]: I think this code is obsoleted, like you originally said.

File: zebra-network/src/peer/error.rs
Comment (teor2345) [inline zebra-network/src/peer/error.rs]: I think `client` could be confusing here, because we also have a `zebra-client` crate. But it's a hard one to name.

Maybe we want to say something like:
```suggestion
    /// The local node closed the connection.
    #[error("Internal node dropped connection")]
    ConnectionDropped,
```

**Review by teor2345 — CHANGES_REQUESTED** [review]
This looks good.

If you want to chat about how we handle pending requests on error, just let me know.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs]: I'm going to look into this today, and see what we might be able to fix here.

Comment (teor2345) [reply zebra-network/src/peer/connection.rs]: See 4c75fa27ab7ddc87d052d4ee5adc8ed9e254ceec, where I:
- Add an `ExitClient` transition, used when the internal client channel is closed or dropped, and there are no more pending requests
- Ignore pending requests after an `ExitClient` transition, because they are impossible
- Reject pending requests when the peer has caused an error (the `Exit` and `ExitRequest` transitions)
- Remove `PeerError::ConnectionDropped`, because it is now handled by `ExitClient`. (Which is an internal error, not a peer error.)

File: zebra-network/src/peer/error.rs
Comment (teor2345) [reply zebra-network/src/peer/error.rs]: Deleted by 4c75fa2

**Comment by teor2345** [comment]
I think we're ready for a review here, but we might want to do some doc updates before we merge this PR.

**Review by teor2345 — CHANGES_REQUESTED** [review]
I think we're pretty good here now, but I'd like to make sure:
- [x] @yaahc checks the commits I added
- [x] we've revised the module docs and other comments as needed

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:688]: Nice

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:10]: We've dealt with the undocumented state, but there are still quite a few undocumented assumptions and invariants. See #1515 for details.

```suggestion
//!
//! This module contains a lot of undocumented assumptions about the Zcash
//! network protocol. We don't know if these assumptions match the `zcashd`
//! implementation. It should be refactored into a cleaner set of
//! request/response pairs (#1515).

```

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:375]: 👍 

These details might be worth linking/moving to the top of the module. But that shouldn't block the PR.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs]: This match statement would be a lot easier to read if we mapped the `Either<Message, Either<Timeout, CancelChannel>>` to a custom enum with `Message`, `Timeout`, and `CancelChannel` variants.

But we should do that in a follow-up PR - it's not a critical change, because it isn't causing any panics.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:510]: Nice cleanup!

**Review by teor2345 — DISMISSED** [review]
Looks good, there are some comments we can tweak, but we can make those changes later.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:10]: This change shouldn't block merging.

Feel free to do it later, or commit and then admin-merge.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:375]: Feel free to do this change later, or commit and then admin-merge.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs]: We also want a `select3` in `peer_set::initialize`, so we should do those changes together:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-network/src/peer_set/initialize.rs#L310

This is now #1783.

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [reply zebra-network/src/peer/connection.rs:375]: I don't know, its pretty specifically tied to this code, my feeling is that it's best to keep it near the logic it is describing, since this is implementation details and not an explanation of how to interact with it

Comment (teor2345) [reply zebra-network/src/peer/connection.rs:375]: Great that works for me

**Review by teor2345 — APPROVED** [review]
The comment I suggested was applied.

**Comment by yaahc** [comment]
All the failures here seem to be on large checkpoint sync for testnet, which is unreliable iirc. cc @teor2345 but I think this is ready for merging.

**Comment by teor2345** [comment]
Yeah we just merged a PR which disabled that test, and it was the only one that failed.

File: zebra-network/src/peer/client.rs
Comment (teor2345) [inline zebra-network/src/peer/client.rs:205]: This change replaces `error_slot` with `PeerError::ConnectionClosed`.

What happened to the error value that used to be here?
Are we hiding more specific errors as part of this change?

File: zebra-network/src/peer/client.rs
Comment (teor2345) [inline zebra-network/src/peer/client.rs:230]: What happens to the `ClientRequest.tx` we used to send on here?
What happens to the specific error in the `error_slot`?

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:371]: What happened to this loop?

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:412]: When does the timer start and finish, compared with the old timer?

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:439]: Should we be awaiting a request here?

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:458]:
```rust
// Special case: ping timeouts fail the connection.
```

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:459]:
```rust
// Other request timeouts fail the request.
```

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:843]: This early exit is probably the right thing to do, but let's check.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:913]: This looping `fail_with` could be the source of the `fail_with` panics.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:850]: Similarly, the early exit is probably correct here, and the loop probably caused those panics.

**Review by teor2345 — COMMENTED** [review]
@yaahc here are my questions about this PR and the hang/slowness in #1801.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:458]: (Let's add these comments back.)

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [inline zebra-network/src/peer/connection.rs:544]: Ah right the loop went here. That's ok.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:371]: It moved into the `run` function.

File: zebra-network/src/peer/connection.rs
Comment (teor2345) [reply zebra-network/src/peer/connection.rs:843]: In this code, a peer has asked us for a list of transactions, and we're sending them back one by one.
If the peer connection fails, we stop sending transactions.
This is the correct behaviour, because we don't send messages on failed connections.

File: zebra-network/src/peer/client.rs
Comment (yaahc) [reply zebra-network/src/peer/client.rs:205]: possibly, but I think the errors we used to grab here were misleading and shouldn't be propagated back to this source. The error slot used to store the error that brought down a `Connection`, but that error is actually associated with one of the client requests created by `call`. We would propagate the error back to that caller over the channel and also copy it to be shared with all future callers. My instinct is that this isn't particularly useful and that we will correctly report all these errors still, they just wont end up being reported many extra times. Imo its fine for subsequent attempts to use a `Client` with a dead `Connection` to just say "sorry this is already closed".

File: zebra-network/src/peer/client.rs
Comment (yaahc) [reply zebra-network/src/peer/client.rs:230]: It's never sent to the `Connection` in the background so we never poll the `rx` so there's no reason to send an error through the `tx` before dropping it. I think this originally got added when we were figuring out the `MustUseSender`, and this was added to avoid panicking because the sender hadn't been used, but we don't even convert to a `MustUseSender` yet here so it should be fine.

There error slot here is gone by the same logic above, the error from a previous request doesn't need to be propagated back to subsequent failed attempts at new requests.

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [reply zebra-network/src/peer/connection.rs:439]: I don't think so, I think I initially set it to `AwaitRequest` durring the refactor but that was incorrect and caused panics, so I commented this out when I was fixing it. Just need to go back and delete this commented out code.

File: zebra-network/src/peer/connection.rs
Comment (yaahc) [reply zebra-network/src/peer/connection.rs:412]: The timer should be constructed at the end of a state transition, same as before. It used to happen in the body of `handle_client_request`, once we've finished processing the request and manually updated the various bits of state to approximate a state transition. Now it is handled in the `TryFrom` impl where we take the `Transition` and use it to construct the subsequent `State`, as far as I can tell this should result in it starting and finishing at the same time.

**Pull request merged**
